### PR TITLE
Cypress with Nginx

### PR DESCRIPTION
**Pull request by coldlink**
## What does this change

So we can run cypress tests against https://profile.thegulocal.com

Previously cypress tests on both mocked and ete tests ran on localhost. This was okay when doing development again identity api as we did not need to rely on cookies set on particular domains.

However as we integrate Okta into our platform for authentication and authorization, this relies on cookies being set on a given domain, i.e the `profile` subdomain. The tests we had in place were workarounds to this problem, and were not testing an actual user scenario, which caused  issues with sign in and sign out for example where cookies were being set incorrectly in the test environment.

## Changes

Updates all of our Cypress Github actions with an additional step to set up the OS environment.

This script installs any OS required packages, downloads [mkcert](https://github.com/FiloSottile/mkcert) and [dev-nginx](https://github.com/guardian/dev-nginx),  installs these packages in the correct place, and sets up the SSL and Nginx config to run tests on `profie.thegulocal.com`.

To facilitate this an [`nginx.conf`](https://github.com/guardian/gateway/blob/b6770d05a8a53ba00689585974b25db82e028760/scripts/nginx.conf) is now included in the project itself to be used specifically by the github actions. This replaces the default `nginx.conf` installed by nginx, and includes a slimmed down version of the `profile.thegulocal.com` configuration from [identity-platform nginx config](https://github.com/guardian/identity-platform/blob/main/nginx/identity.conf).

We then update the environment variables and github actions configuration to point to the `profile.thegulocal.com` baseUri, and associated changes, e.g. the health check.

The local cypress configuration is also then changed to make sure that requires nginx and profile.thegulocal.com to be set up locally to be able to run the tests, which includes change the bash scripts, and updating/fixing any tests which broke as a result of this change. This included the removal and major reworking of some specific tests.

Finally a change was made to add `DevProfileIdClient` to `performAuthorizationCodeFlow` method which is required to be able to use `profile.thegulocal.com` through Oauth/OpenId Connect flows as we're unable to set up the custom domain in our Okta enviroment. This change was initially introduced in https://github.com/guardian/gateway/pull/1558 and expanded to the method in this PR.

Also added `retryOnStatusCodeFailure` for `createTestUser` to potentially avoid flaky tests relying on this functionality.

## Related PR

https://github.com/guardian/identity-platform/pull/511

## Follow up TODO
- Move more of the ete tests to okta too
- Review all of our current tests
  - There are somethings in the ete tests that would be served better as mocked tests and vice versa
  - Guidelines about where a given test should go

### CODE REVIEW

**Review by rebecca-thompson — COMMENTED** [review]
Just a few comments, but otherwise looking great!

File: .github/workflows/cypress-ete-okta.yml [hunk 1]
```diff
@@ -67,11 +84,18 @@ jobs:
           REDIS_PASSWORD: ${{ secrets.REDIS_PASSWORD }}
           REDIS_SSL_ON: false
           RATE_LIMITER_CONFIG: '{"enabled":true,"defaultBuckets":{"globalBucket":{"capacity":500,"addTokenMs":50},"ipBucket":{"capacity":100,"addTokenMs":50},"emailBucket":{"capacity":100,"addTokenMs":50},"oktaIdentifierBucket":{"capacity":100,"addTokenMs":50},"accessTokenBucket":{"capacity":4,"addTokenMs":5000}},"routeBuckets":{}}'
+          # Required for the Cypress tests to run as we're unable to verify the created certs
+          NODE_TLS_REJECT_UNAUTHORIZED: 0
```
[the remark this answers is not in the document]
Comment (coldlink) [reply .github/workflows/cypress-ete-okta.yml:88]: Locally you can do the following if you followed the [nginx setup guide](https://github.com/guardian/gateway/blob/main/docs/setup.md#nginx). It was not possible to use this same method in Github Actions due to user permissions on the OS.

```
# in bash/zsh config
export NODE_EXTRA_CA_CERTS="$(mkcert -CAROOT)/rootCA.pem"
```

File: .github/workflows/cypress-ete.yml [hunk 2]
```diff
@@ -28,23 +28,39 @@ jobs:
         with:
           node-version-file: '.nvmrc'
           cache: 'yarn'
+      - name: Setup OS, Nginx, and Certs
+        run: |
```
[the remark this answers is not in the document]
Comment (coldlink) [reply .github/workflows/cypress-ete.yml:32]: Unfortunately not, gha doesn't let you make a reusable steps in each action.

there is a concept of [reusing workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows), but this require rearchitecting all our actions, which can be a future improving, but will likely take a lot more work

File: README.md [hunk 3]
```diff
@@ -33,20 +33,6 @@ $ make dev
...
-```sh
-$ docker-compose up -d
-```
-
-4. Logs:
-
-```sh
-$ docker-compose logs -f
-```
-
```
[the remark this answers is not in the document]
Comment (coldlink) [reply README.md:49]: Yeah it does, but I decided to remove it from the main readme, as it suggests that we recommend using docker to run the project, which is not the case, and instead mention in the setup readme that if someone prefers to use docker they can.

File: .github/workflows/cypress-ete-okta.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (coldlink) [reply .github/workflows/cypress-ete-okta.yml:88]: Which scripts are you talking about?

This `NODE_EXTRA_CA_CERTS` is required for using https://profile.thegulocal.com even when developing locally, not just cypress, 

I think it's a super small step to do, and mkcert itself doesn't do it automatically (which is where the solution is from): https://github.com/FiloSottile/mkcert#using-the-root-with-nodejs

File: src/server/lib/okta/oauth.ts [hunk 4]
```diff
@@ -19,10 +25,17 @@ import {
...
+  req: Request,
   res: ResponseWithRequestState,
   sessionToken?: string,
   confirmationPagePath?: RoutePaths,
 ) => {
+  // Determine which OpenIdClient to use, in DEV we use the DevProfileIdClient, otherwise we use the ProfileOpenIdClient
+  const OpenIdClient =
+    stage === 'DEV' && req.get('X-GU-Okta-Env')
+      ? DevProfileIdClient(req.get('X-GU-Okta-Env') as string)
+      : ProfileOpenIdClient;
```
[the remark this answers is not in the document]
Comment (coldlink) [reply src/server/lib/okta/oauth.ts:37]: I can see your point, and I tried doing this in the initial PR where this was set up (https://github.com/guardian/gateway/pull/1558).

The header is required, as it's the only way that nginx can tell gateway where the okta request originated from: https://github.com/guardian/identity-platform/blob/35ec0e05e17fa1c03bf7feed279a810b3c841749/nginx/identity.conf#L40

As nginx is determining the routing and not gateway, we can't use configuration in gateway to determine this.

I'll add some more to the readme about the `gu_okta_env` cookie, which is only read in nginx to determine where okta should point to, and is optional, as nginx points to CODE by default.

[the remark this answers is not in the document]
Comment (coldlink) [reply src/server/lib/okta/oauth.ts:37]: I can try abstracting this away to, into something like a `getOpenIdClient` function

File: cypress/integration/ete/sign_out/sign_out.spec.ts [hunk 5]
```diff
@@ -9,42 +9,52 @@ describe('Sign out flow', () => {
...
           cy.getCookie(cookie).should('exist');
         });
 
-        cy.request('/signout');
+        // attempt to sign out and redirect to reset password as it's on same domain
+        const postSignOutReturnUrl = `https://${Cypress.env(
+          'BASE_URI',
+        )}/reset-password`;
+        cy.visit(
+          `/signout?useOkta=true&returnUrl=${encodeURIComponent(
```
[the remark this answers is not in the document]
Comment (coldlink) [reply cypress/integration/ete/sign_out/sign_out.spec.ts]: Good spot! I've removed it.

File: docs/setup.md [hunk 6]
```diff
@@ -86,11 +74,17 @@ Here's a quick-start configuration that you can copy-paste into `.ratelimit.json
 
 When you start Gateway it will first attempt to read from the `RATE_LIMITER_CONFIG` environment variable. If that is not defined, it will read the `.ratelimit.json` file we just made into its configuration.
 
+### Environment File
+
+Populate `.env` files by using the examples from [`.env.example`](../.env.example). The `.env` files should **never** be committed as they store secrets.
+
+Depending on which stage (`DEV` or `CODE`) you want to connect to [Identity API (IDAPI)](https://github.com/guardian/identity), the `IDAPI_CLIENT_ACCESS_TOKEN` and `IDAPI_BASE_URL` variables will be different. If using the S3 config, it will point to the `CODE` instance of IDAPI.
+
```
[the remark this answers is not in the document]
Comment (coldlink) [reply docs/setup.md:82]: https://github.com/guardian/gateway/pull/1588/commits/954248a105d5f92901a948f9aca60d14e7a10361 see this commit

File: cypress/integration/mocked/change_password.spec.ts [hunk 7]
```diff
@@ -127,10 +127,6 @@ describe('Password change flow', () => {
       cy.contains('use a password that is hard to guess');
       cy.get('button[type="submit"]').click().should('not.be.disabled');
       cy.contains('Please use a password that is hard to guess.');
-      cy.get('input[name="password"]').type('iamaveryuniqueandlongstring');
-      cy.wait('@breachCheck');
-      cy.contains('Valid password');
-      cy.get('button[type="submit"]').click().should('be.disabled');
```
[the remark this answers is not in the document]
Comment (coldlink) [reply cypress/integration/mocked/change_password.spec.ts:133]: Added the checks to the other test!

File: .github/workflows/cypress-ete-okta.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (rebecca-thompson) [reply .github/workflows/cypress-ete-okta.yml:88]: I was referring to the scripts to run the cypress tests in the root of the project `cypress-mocked.sh` and `cypress-ete.sh`

**Review by rebecca-thompson — APPROVED** [review]

**Pull request merged**